src/util/key.rs

```diff
@@ -11,86 +11,94 @@
 // If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 //
 
-//! Private key
+//! Bitcoin Keys
 //!
-//! A private key represents the secret data associated with its proposed use
+//! Structures and methods representing the public/secret data associated with
+//! its proposed use
 //!
 
 use std::fmt::{self, Write};
 use std::str::FromStr;
 use secp256k1::{self, Secp256k1};
-use secp256k1::key::{PublicKey, SecretKey};
 use util::address::Address;
 use consensus::encode;
 use network::constants::Network;
 use util::base58;
 
-#[derive(Clone, PartialEq, Eq)]
-/// A Bitcoin ECDSA private key
-pub struct Privkey {
-    /// Whether this private key represents a compressed address
+/// A Bitcoin ECDSA public key
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PublicKey {
+    /// Whether this public key represents a compressed address
     pub compressed: bool,
-    /// The network on which this key should be used
-    pub network: Network,
     /// The actual ECDSA key
-    pub key: SecretKey
+    pub key: secp256k1::PublicKey,
 }
 
-impl Privkey {
-    /// Creates a `Privkey` from a raw secp256k1 secret key
-    #[inline]
-    pub fn from_secret_key(key: SecretKey, compressed: bool, network: Network) -> Privkey {
-        Privkey {
+impl PublicKey {
+    /// serialize it into a vec
+    pub fn serialize(&self) -> Vec<u8> {
+        if self.compressed {
+            self.key.serialize().to_vec()
```

Comment:
Using arrays in interfaces is kinda awkward. I know @apoelstra likes it, but I think it would make a lot more sense for `secp256k1::key::PublicKey::serialize` to return a `Vec<u8>` instead of an array. I don't think there is a way to convert an array into a Vec without reallocating, right?

Reply:
Yeah... I'm not sure how to best handle this... @apoelstra?

Reply:
rust-secp256k1 is not going to return `Vec<u8>`s for anything unless it absolutely has to. (I think that `Signature::serialize_der` does, but that's it...and it was an `ArrayVec` for a while until I got concerned about extra deps for basic datatypes).

We could add methods that can serialize in/out of an `io::Read` and `io::Write` I suppose. It does not seem reasonable to me that a serialization method would force the user to allocate.

Reply:
@apoelstra I think we should do `io::Read` and `io::Write` route.

Reply:
@apoelstra Done: https://github.com/rust-bitcoin/rust-secp256k1/issues/76